src/basic/utf8.c

```diff
@@ -314,18 +314,25 @@ size_t utf8_encode_unichar(char *out_utf8, char32_t g) {
         return 0;
 }
 
-char *utf16_to_utf8(const void *s, size_t length) {
+char *utf16_to_utf8(const char16_t *s, size_t length /* bytes! */) {
         const uint8_t *f;
         char *r, *t;
 
-        r = new(char, (length * 4 + 1) / 2 + 1);
+        assert(s);
+
+        /* Input length is in bytes, i.e. the shortest possible character takes 2 bytes. Each unicode character may
+         * take up to 4 bytes in UTF-8. Let's also account for a trailing NUL byte. */
+        if (length * 2 < length)
```

Comment:
Not really, no. Only the multiplication by two can overflow. After we multiplied we know the lowest bit is off, hence the increment by one will always work...

Generally I think it's wise to check for overflows individually, rather than trying to be smart and converting them into a single test, simply because this all so fragile and the tests are hard to grok in their full effect